### ISSUE TITLE
appcenter_upload doesn't fail on error

### ISSUE THREAD

**Issue by jenskuhrjorgensen** [microsoft/fastlane-plugin-appcenter#327]
Hi

When an error happens during upload and distribution of an iOS app via `appcenter_upload` fails, e.g. due to an internal server error, the upload doesn't fail meaning the distribution error is not caught in our pipelines and nobody knows that something went wrong before our testers report that the app update is missing. Here is an example of the issue where an app is uploaded to AppCenter and then distributed to two distribution groups. Distributing to the first one fails, the second one succeeds and then the entire `appcenter_upload` command is marked as successful:

```
[14:43:45]: ------------------------------
[14:43:45]: --- Step: appcenter_upload ---
[14:43:45]: ------------------------------
[14:43:45]: Starting release upload...
[14:43:46]: Setting Metadata...
[14:43:46]: Metadata set
[14:43:46]: Uploading release binary...
[14:43:49]: Chunk uploaded
[14:43:51]: Chunk uploaded
[14:43:53]: Chunk uploaded
[14:43:55]: Chunk uploaded
[14:43:57]: Chunk uploaded
[14:43:59]: Chunk uploaded
[14:44:01]: Chunk uploaded
[14:44:03]: Chunk uploaded
[14:44:07]: Chunk uploaded
[14:44:09]: Chunk uploaded
[14:44:11]: Chunk uploaded
[14:44:13]: Chunk uploaded
[14:44:15]: Chunk uploaded
[14:44:15]: Binary uploaded
[14:44:15]: Finishing release...
[14:44:16]: Upload finished
[14:44:16]: Waiting for release to be ready...
[14:44:25]: Release '225' committed: <REDACTED>
[14:44:26]: Release '225' (2023.14) was successfully updated
[14:44:27]: Error adding to group 500: {"code"=>"internal_server_error", "message"=>"internal error (Support ID: 42cb064f-fd66-4b1f-b6d8-7d75a02dbc16)"}
[14:44:27]: Release '225' was not found for destination '<REDACTED>'
[14:44:29]: Release '225' (2023.14) was successfully distributed'
[14:44:29]: Release '225' (2023.14) was successfully distributed to group "<REDACTED>"
[14:44:29]: Release '225' is available for download at: <REDACTED>
[14:44:29]: dSYM is found
[14:44:29]: Successfully uploaded app to AppCenter. Download via <REDACTED>.
```

I would expect the upload to fail if something goes wrong, but instead "Successfully uploaded app to AppCenter" is printed and all our pipeline lamps are green.

We've reported the issue behind the 500 error to the AppCenter support for them to look into, but though the issue with the error handling in the Fastlane plugin was more fit in a bug report in here.

We're running `fastlane-plugin-appcenter (2.1.0)`.

**Comment by DmitriyKirakosyan** [comment]
@jenskuhrjorgensen, thanks for reaching out! This exception suggests a backend issue. Our support team will assist you.

**Comment by jenskuhrjorgensen** [comment]
Hi @DmitriyKirakosyan
Yes, det root cause for this failure is probably a backend issue (and I already reported that to your support), but I would expect the `appcenter_upload` command to fail instead of just continuing as if all is good, when the backend throws a 500 back. Wouldn't you say so too?

**Comment by DmitriyKirakosyan** [comment]
It's difficult to say if the uploading should stop if we fail adding an app to a group. This action is not a blocker for the CodePush release and can be performed manually on Portal.

**Comment by jenskuhrjorgensen** [comment]
Yes, it can be performed manually on the portal, but I would rather replay a failing pipeline than have to find my way to the portal and add it there manually. Doing it the manual way defeats the purpose of automated ci/cd and requires distributed knowledge of how to navigate the AapCenter portal. We are very rarely visiting the AC portal because we don't have to - as long as our pipelines deliver and distribute our apps to our internal testers all is good. But whenever we discover that automatic distribution has failed we have probably already wasted hours because testers have not tested the right version or because they have struggled with trying to get the newest version - which has bot been correctly distributed.

I don't see how it is relevant that code push is possible even if distribution fails. If the upload doesn't distribute to the groups that you specify and the backend throws a an error, that should result in an error in the frontend too - because the command didn't successfully do what you expected it to do.

**Comment by DmitriyKirakosyan** [comment]
You make a valid point. I understand the importance of seamless CI/CD processes for you.

However, the release should actually be uploaded according to your logs. The only missing step is adding a distribution group to the application, which is normally required once per app. This only determines if specific testers can download the release.

Can you confirm that this is what happens on your side?
If yes, I have an idea. We can introduce a separate argument which would decide whether we fail the release if adding distribution group fails. What do you think?

**Comment by jenskuhrjorgensen** [comment]
The release is uploaded but it is not distributed to both distribution groups. In the example logs from the OP the distribution to the first distribution group fails whereas the distribution to the second distribution group succeeds. And yes, this exactly means that some testers cannot download the app/are not notified of the new app version because it has not been made available to them in the distribution group they using.

If you are suggesting a parameter ala "fail_on_distribution_error" (or maybe "**ignore_distribution_errors**" would be more appropriate because I think it should by default fail on error but if someone don't care about and/or experience a lot of distribution errors then they can choose to ignore those by actively setting this parameter) that would be fine for us. As long as we can make it fail on error I think we are golden! :)

**Comment by jenskuhrjorgensen** [comment]
Hi @DmitriyKirakosyan 
I saw your reaction to my comment but do you have a comment to my comment or are you already busy implementing the feature? :D

**Comment by lucen-ms** [comment]
@jenskuhrjorgensen at this moment it works as designed - the task fails when upload fails but not when adding to distribution group fails. We do not plan to change this behavior.

**Comment by jenskuhrjorgensen** [comment]
Hi @lucen-ms 

Can you elaborate on how it is considered "success" when it isn't all the tasks that the command handles that are succeeding? Is there a separate "distribute" command that I can use if I want to make sure that distribution succeeds and if I want to have my pipeline fail if distribution fails? Right now it feels like you don't really care if distribution succeeds or not - sometimes it works sometimes it doesn't - and I'm not notified whether I was lucky or not.
